contracts/bridge/contracts/common/message/OutboundLane.sol

```diff
@@ -0,0 +1,332 @@
+        key = (thisChainPosition << 224) + (bridgedChainPosition << 192) + (lanePosition << 128) + nonce;
+    }
+
+    function message_size() public view returns (uint256 size) {
+        size = outboundLaneNonce.latest_generated_nonce - outboundLaneNonce.latest_received_nonce;
+    }
+
+	/// Returns saved outbound message payload.
+    function message(uint256 nonce) public view returns (MessageData memory) {
+        return messages[encodeMessageKey(nonce)];
+    }
+
+	// Get lane data from the storage.
+    function data() public view returns (OutboundLaneData memory lane_data) {
+        uint256 size = message_size();
+        if (size > 0) {
+            lane_data.messages = new Message[](size);
+            uint256 begin = outboundLaneNonce.latest_received_nonce + 1;
+            for (uint256 index = 0; index < size; index++) {
+                uint256 nonce = index + begin;
+                uint256 key = encodeMessageKey(nonce);
+                lane_data.messages[index] = Message(MessageKey(thisChainPosition, bridgedChainPosition, lanePosition, nonce), messages[key]);
+            }
+        }
+        lane_data.latest_received_nonce = outboundLaneNonce.latest_received_nonce;
+    }
+
+    /* Private Functions */
+
+	/// commit lane data to the `commitment` storage.
+    function commit() internal returns (bytes32) {
+        commitment = hash(data());
+        return commitment;
+    }
+
+    function extract_substrate_inbound_lane_info(InboundLaneData memory lane_data) internal pure returns (uint256 total_unrewarded_messages, uint256 last_delivered_nonce) {
+        total_unrewarded_messages = lane_data.last_delivered_nonce - lane_data.last_confirmed_nonce;
+        last_delivered_nonce = lane_data.last_delivered_nonce;
+    }
+
+	// Confirm messages delivery.
+    function confirm_delivery(InboundLaneData memory subInboundLaneData) internal returns (DeliveredMessages memory confirmed_messages) {
+        (uint256 total_messages, uint256 latest_delivered_nonce) = extract_substrate_inbound_lane_info(subInboundLaneData);
+        require(total_messages < 256, "Lane: InvalidNumberOfMessages");
+
+        UnrewardedRelayer[] memory relayers = subInboundLaneData.relayers;
+        OutboundLaneNonce memory nonce = outboundLaneNonce;
+        require(latest_delivered_nonce > nonce.latest_received_nonce, "Lane: NoNewConfirmations");
+        require(latest_delivered_nonce <= nonce.latest_generated_nonce, "Lane: FailedToConfirmFutureMessages");
+        // that the relayer has declared correct number of messages that the proof contains (it
+        // is checked outside of the function). But it may happen (but only if this/bridged
+        // chain storage is corrupted, though) that the actual number of confirmed messages if
+        // larger than declared.
+        require(latest_delivered_nonce - nonce.latest_received_nonce <= total_messages, "Lane: TryingToConfirmMoreMessagesThanExpected");
```

Comment:
this condition equals with `require(nonce.latest_received_nonce >= subInboundLaneData.last_confirmed_nonce)`, right?